### ISSUE TITLE
panic: runtime error: invalid memory address or nil pointer dereference

### ISSUE THREAD

**Issue by bitsofinfo** [TykTechnologies/tyk#1011]
**Do you want to request a *feature* or report a *bug*?**

bug

**What is the current behavior?**

My listen path is `/my/api/1.0`, strip listen path is checked

I need to whitelist GET's to `/my/api/1.0?wsdl`

In a GET whitelist I put in `?wsdl` and get this stack trace on a request to the gateway 

In a GET whitelist I put in `/?wsdl` throws no error but doesn't work (get forbidden on a request)

An explicit whitelist on `/my/api/1.0?wsdl` doesn't work either (get forbidden on a request)

```
2017-08-15T20:28:02.413172594Z 2017/08/15 20:28:02 http: panic serving 10.218.8.28:50319: runtime error: invalid memory address or nil pointer dereference
2017-08-15T20:28:02.413177794Z goroutine 245185 [running]:
2017-08-15T20:28:02.413181694Z net/http.(*conn).serve.func1(0xc4409f3280)
2017-08-15T20:28:02.413185594Z 	/usr/local/go/src/net/http/server.go:1491 +0x12a
2017-08-15T20:28:02.413189794Z panic(0xccaae0, 0xc420010070)
2017-08-15T20:28:02.413193793Z 	/usr/local/go/src/runtime/panic.go:458 +0x243
2017-08-15T20:28:02.413197893Z regexp.(*Regexp).get(0x0, 0x19)
2017-08-15T20:28:02.413202093Z 	/usr/local/go/src/regexp/regexp.go:208 +0x26
2017-08-15T20:28:02.413206193Z regexp.(*Regexp).doExecute(0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc441521e60, 0x19, 0x0, 0x0, ...)
2017-08-15T20:28:02.413210193Z 	/usr/local/go/src/regexp/exec.go:416 +0x40
2017-08-15T20:28:02.413213893Z regexp.(*Regexp).MatchString(0x0, 0xc441521e60, 0x19, 0x19)
2017-08-15T20:28:02.413217693Z 	/usr/local/go/src/regexp/regexp.go:416 +0x88
2017-08-15T20:28:02.413221593Z main.(*APISpec).IsURLAllowedAndIgnored(0xc43dd23d60, 0xc440a06ae0, 0x3, 0xc441521d00, 0x19, 0xc441521e20, 0x1, 0xc43fca51f8, 0x556d85, 0x20, ...)
2017-08-15T20:28:02.413225693Z 	/src/github.com/TykTechnologies/tyk/api_definition_manager.go:1066 +0x108
2017-08-15T20:28:02.413235493Z main.(*APISpec).IsRequestValid(0xc43dd23d60, 0xc44026d0e0, 0xc441521e0c, 0xdbe1d3, 0x9, 0x15, 0xc441521e00)
2017-08-15T20:28:02.413239593Z 	/src/github.com/TykTechnologies/tyk/api_definition_manager.go:1291 +0x17d
2017-08-15T20:28:02.413243593Z main.(*VersionCheck).ProcessRequest(0xc440166710, 0x12ba260, 0xc43fee2340, 0xc44026d0e0, 0x0, 0x0, 0x15, 0x0, 0x14)
2017-08-15T20:28:02.413247393Z 	/src/github.com/TykTechnologies/tyk/middleware_version_check.go:51 +0x53
2017-08-15T20:28:02.413251293Z main.CreateMiddleware.func1.1(0x12ba260, 0xc43fee2340, 0xc44026d0e0)
2017-08-15T20:28:02.413255093Z 	/src/github.com/TykTechnologies/tyk/middleware.go:76 +0x637
2017-08-15T20:28:02.413258793Z net/http.HandlerFunc.ServeHTTP(0xc440154680, 0x12ba260, 0xc43fee2340, 0xc44026d0e0)
2017-08-15T20:28:02.413262593Z 	/usr/local/go/src/net/http/server.go:1726 +0x44
2017-08-15T20:28:02.413266293Z main.CreateMiddleware.func1.1(0x12ba260, 0xc43fee2340, 0xc44026d0e0)
2017-08-15T20:28:02.413270193Z 	/src/github.com/TykTechnologies/tyk/middleware.go:100 +0x83c
2017-08-15T20:28:02.413273993Z net/http.HandlerFunc.ServeHTTP(0xc4401546c0, 0x12ba260, 0xc43fee2340, 0xc44026d0e0)
2017-08-15T20:28:02.413298493Z 	/usr/local/go/src/net/http/server.go:1726 +0x44
2017-08-15T20:28:02.413302693Z main.CreateMiddleware.func1.1(0x12ba260, 0xc43fee2340, 0xc44026d0e0)
2017-08-15T20:28:02.413308293Z 	/src/github.com/TykTechnologies/tyk/middleware.go:100 +0x83c
2017-08-15T20:28:02.413312593Z net/http.HandlerFunc.ServeHTTP(0xc440154700, 0x12ba260, 0xc43fee2340, 0xc44026d0e0)
2017-08-15T20:28:02.413316793Z 	/usr/local/go/src/net/http/server.go:1726 +0x44
2017-08-15T20:28:02.413320893Z main.CreateMiddleware.func1.1(0x12ba260, 0xc43fee2340, 0xc44026d0e0)
2017-08-15T20:28:02.413324993Z 	/src/github.com/TykTechnologies/tyk/middleware.go:100 +0x83c
2017-08-15T20:28:02.413329193Z net/http.HandlerFunc.ServeHTTP(0xc440154740, 0x12ba260, 0xc43fee2340, 0xc44026d0e0)
2017-08-15T20:28:02.413333193Z 	/usr/local/go/src/net/http/server.go:1726 +0x44
2017-08-15T20:28:02.413337293Z github.com/TykTechnologies/tyk/vendor/github.com/gorilla/context.ClearHandler.func1(0x12ba260, 0xc43fee2340, 0xc44026d0e0)
2017-08-15T20:28:02.413341493Z 	/src/github.com/TykTechnologies/tyk/vendor/github.com/gorilla/context/context.go:141 +0x8b
2017-08-15T20:28:02.413345593Z net/http.HandlerFunc.ServeHTTP(0xc440169a20, 0x12ba260, 0xc43fee2340, 0xc44026d0e0)
2017-08-15T20:28:02.413349993Z 	/usr/local/go/src/net/http/server.go:1726 +0x44
2017-08-15T20:28:02.413353993Z github.com/TykTechnologies/tyk/vendor/github.com/gorilla/mux.(*Router).ServeHTTP(0xc43fa20d20, 0x12ba260, 0xc43fee2340, 0xc44026d0e0)
2017-08-15T20:28:02.413358293Z 	/src/github.com/TykTechnologies/tyk/vendor/github.com/gorilla/mux/mux.go:114 +0x10d
2017-08-15T20:28:02.413362293Z net/http.(*ServeMux).ServeHTTP(0xc440262b10, 0x12ba260, 0xc43fee2340, 0xc440fb40f0)
2017-08-15T20:28:02.413369093Z 	/usr/local/go/src/net/http/server.go:2022 +0x7f
2017-08-15T20:28:02.413373393Z net/http.serverHandler.ServeHTTP(0xc422dfd300, 0x12ba260, 0xc43fee2340, 0xc440fb40f0)
2017-08-15T20:28:02.413377493Z 	/usr/local/go/src/net/http/server.go:2202 +0x7d
2017-08-15T20:28:02.413381493Z net/http.(*conn).serve(0xc4409f3280, 0x12bb2e0, 0xc4409b3ac0)
2017-08-15T20:28:02.413385493Z 	/usr/local/go/src/net/http/server.go:1579 +0x4b7
2017-08-15T20:28:02.413389492Z created by net/http.(*Server).Serve
2017-08-15T20:28:02.413393492Z 	/usr/local/go/src/net/http/server.go:2293 +0x44d
```

**What is the expected behavior?**

Should not throw such an error, if this is an invalid whitelist match at the root of a uri path, should be caught by the gateway api for configuring a whitelist?

**Which versions of Tyk affected by this issue? Did this work in previous versions of Tyk?**

2.3.7/1.3.7

**Comment by mvdan** [comment]
Thanks for the report - will look into this straight away.

Could you provide your gateway config and the API definition you're using?

**Comment by mvdan** [comment]
Okay, so the issue here is simpler than I thought.

The cause is the regex `?wsdl` - it's not a valid regex. If you try this with Tyk built from master, it will error on startup with:

    panic: regexp: Compile(`?wsdl`): error parsing regexp: missing argument to repetition operator: `?`

In other words, you probably want `\?wsdl`.

The reason why 2.3.x panics with a more confusing message is because the stable version sometimes doesn't check for errors in user-provided regexes. The upcoming 2.4 version does, hence the immediate crash with a proper error message.

I'm going to close this, as this is an issue in the API definition and the confusing panic has been addressed in the upcoming 2.4. Thanks!
